File: src/lib/ai/providers/openai.ts
import OpenAI from 'openai';
import { AIProviderError } from '@/lib/errors';
import { env } from '@/lib/server-env';

let client: OpenAI | null = null;

function getClient() {
  const apiKey = env.OPENAI_API_KEY;
  if (!apiKey?.trim()) {
    throw new AIProviderError('OPENAI_API_KEY is not configured', 'openai');
  }

  client ??= new OpenAI({ apiKey });
  return client;
}

const MAX_OUTPUT_TOKENS = 8192;

export async function generateObject<T>(prompt: string, schema: object, model = env.OPENAI_MODEL): Promise<T> {
  const res = await getClient().chat.completions.create({
    model,
    max_tokens: MAX_OUTPUT_TOKENS,
    messages: [{ role: 'user', content: prompt }],
    tools: [{
      type: 'function',
      function: {
        name: 'output',
        description: 'Return structured output',
        parameters: schema as Record<string, unknown>,
      },
    }],
    tool_choice: { type: 'function', function: { name: 'output' } },
  });

  const choice = res.choices[0];
  const finishReason = choice?.finish_reason;

  if (finishReason === 'length') {
    throw new AIProviderError(
      `OpenAI response truncated at ${MAX_OUTPUT_TOKENS} tokens (finish_reason=length)`,
      'openai',
      undefined,
      true,
    );
  }

  const toolCall = choice?.message.tool_calls?.[0];
  if (!toolCall || toolCall.type !== 'function') {
    throw new AIProviderError('No function call in OpenAI response', 'openai');
  }

  try {
    return JSON.parse(toolCall.function.arguments) as T;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new AIProviderError(
      `Failed to parse OpenAI tool arguments (finish_reason=${finishReason}): ${msg}`,
      'openai',
      undefined,
      true,
    );
  }
}

File: services/api/app/ai_clients/openai.py
from __future__ import annotations

from collections.abc import AsyncIterator

from openai import AsyncOpenAI

from ..config import get_settings
from ..errors import AIServiceError
from .base import await_provider, parse_json_response

PROVIDER = "openai"


async def generate_json(messages: list[dict], schema_hint: str) -> dict:
    settings = get_settings()
    timeout = settings.ai_request_timeout_seconds
    if not settings.openai_api_key:
        raise AIServiceError("OPENAI_API_KEY is not configured", PROVIDER)

    client = AsyncOpenAI(api_key=settings.openai_api_key, timeout=timeout)
    response = await await_provider(
        client.chat.completions.create(
            model=settings.openai_model,
            messages=[
                {"role": "system", "content": f"You are a JSON API. Respond with valid JSON only. Use this exact schema:\n{schema_hint}"},
                *[{"role": msg["role"], "content": msg["content"]} for msg in messages],
            ],
            response_format={"type": "json_object"},
        ),
        PROVIDER,
        timeout + 5,
        "OpenAI request",
    )
    content = response.choices[0].message.content or "{}"
    return parse_json_response(content, PROVIDER)


async def get_web_context(prompt: str) -> str:
    settings = get_settings()
    if not settings.openai_api_key:
        return ""

    client = AsyncOpenAI(api_key=settings.openai_api_key, timeout=settings.ai_request_timeout_seconds)
    try:
        response = await await_provider(
            client.responses.create(
                model=settings.openai_web_search_model or settings.openai_model,
                tools=[{"type": "web_search_preview"}],
                tool_choice="auto",
                input=prompt,
            ),
            PROVIDER,
            settings.ai_request_timeout_seconds + 15,
            "OpenAI web search request",
        )
    except AIServiceError:
        return ""
    return (getattr(response, "output_text", "") or "").strip()


async def stream_reply(system_prompt: str, messages: list[dict]) -> AsyncIterator[str]:
    settings = get_settings()
    timeout = settings.ai_request_timeout_seconds
    if not settings.openai_api_key:
        raise AIServiceError("OPENAI_API_KEY is not configured", PROVIDER)

    client = AsyncOpenAI(api_key=settings.openai_api_key, timeout=timeout)
    stream = await client.chat.completions.create(
        model=settings.openai_model,
        stream=True,
        messages=[{"role": "system", "content": system_prompt}, *messages],
    )
    async for chunk in stream:
        text = chunk.choices[0].delta.content or ""
        if text:
            yield text
